Frontend/feat/error boundary

thdk opened the pull request:
Let me know if tests would be needed for some part of this PR.

I didn't find an api endpoint for logging errors to our backend, is that correct?

thdk committed:
chore: add react-error-boundary to our dependencies

thdk committed:
feat: add ErrorFallback component

This component can be displayed by our ErrorBoundary component in case an error happens. BugReport component was changed so it can be used in a UI where there is always enough space to display the 'Report a problem' label.

thdk committed:
feat: add ErrorBoundary component

This component can be used to implement error reporting to our backend service. It has a default fallback component set but might be needed to override this default for specific boundaries (such as boundary for our map?).

In the future we might add a required prop to this ErrorBoundary component to specifya 'location' of our app so errors can be grouped by their location. Eg. request form, map, edit profile, ...

thdk committed:
feat: wrap each route with our ErrorBoundary

Note that Navigation component isn't wrapped with ErrorBoundary. Can be added later?

thdk committed:
style: fix formatting issue on profile page

thdk committed:
test: add a test for isResponsive prop of BugReport component

thdk committed:
Merge branch 'develop' into frontend/feat/error-boundary

lucaslcode commented on app/frontend/package.json:
```diff
@@ -16,6 +16,7 @@
     "maplibre-gl": "^1.14.0",
     "react": "^17.0.0",
     "react-dom": "^17.0.0",
+    "react-error-boundary": "^3.1.1",
```
I'm debating with myself if it's worth adding a dependency for this.

The pro is our code can be cleaner, but the con is we are then relying on this package being maintained. Considering we are hoping to support the app long-term, we shouldn't add dependencies too lightly, and it's not that difficult to manually do an [error boundary component](https://reactjs.org/docs/error-boundaries.html).

What do you think?

thdk replied:
It's a small dependency with many weekly downloads and no other dependencies except for @babel/runtime which I raised a question for on the react-error-boundary repo to know why they have that dependency.
Making your own error boundary is a possibility and is fairly simple indeed but doesn't mean it won't need maintenance in the future. I would say that, if necessary, we can make the ErrorBoundary ourselves once we notice problems with maintenance of react-error-boundary.

lucaslcode commented on app/frontend/src/AppRoute.tsx:
```diff
@@ -46,6 +47,11 @@ export default function AppRoute({
 
   const classes = useStyles();
 
+  const RouteContent = useMemo(
+    () => () => <ErrorBoundary>{children}</ErrorBoundary>,
+    [children]
+  );
```
I don't think there's any benefit to memoizing this.

lucaslcode replied:
I see, you switched to `component` instead of `render` below. If you do `render={() => <RouteContent>{children}</RouteContent>`, you won't need to memoize.

thdk replied:
Ok I drop the useMemo. 
`const RouteContent = () => <ErrorBoundary>{children}</ErrorBoundary>;`

Not sure if I understand though if there is a difference between:
`<Route {...otherProps} render={() => <RouteContent />} />`
and
`<Route {...otherProps} component={RouteContent} />`

lucaslcode commented on app/frontend/src/components/ErrorBoundary/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,17 @@
+
+export default function ErrorBoundary({
+  children,
+}: PropsWithChildren<unknown>) {
```
`ReactNode` is the correct type for children usually.

thdk replied:
I've learned to use PropsWithChildren because the type of children can change internally in React (unlikely I know) and this code would still have correct typings. Both of them are correct to my knowledge.

If you prefere `children: ReactNode`, that's ok :)

thdk replied:
except that when using PropsWithChildren type, the children property is optional while here for ErrorBoundary component the children are required.

lucaslcode commented on app/frontend/src/components/ErrorBoundary/ErrorBoundary.tsx:
```diff
@@ -0,0 +1,17 @@
+export default function ErrorBoundary({
+  children,
+}: PropsWithChildren<unknown>) {
+  const onError = useCallback(() => {
+    // TODO: log error to backend?
+  }, []);
```
remove for now, we can put it in when we have proper error reporting

lucaslcode commented on app/frontend/src/components/ErrorFallback/ErrorFallback.tsx:
```diff
@@ -0,0 +1,51 @@
+}));
+
+export default function ErrorFallback() {
+  const { actions, report } = useStyles();
```
We prefer not destructuring for `classes`

lucaslcode commented on app/frontend/src/components/ErrorFallback/ErrorFallback.tsx:
```diff
@@ -0,0 +1,51 @@
+
+  const history = useHistory();
+
+  const handleRefresh = useCallback(() => history.go(0), [history]);
```
No need to memoize simple functions

lucaslcode commented on app/frontend/src/components/ErrorFallback/ErrorFallback.tsx:
```diff
@@ -0,0 +1,51 @@
+    <>
+      <PageTitle>{ERROR_HEADING}</PageTitle>
+      <Typography variant="body1">{ERROR_INFO}</Typography>
+      <Box className={report}>
```
We used to use `Box` and there are probably some lying around, but we use `div` now

lucaslcode commented on app/frontend/src/components/ErrorFallback/ErrorFallback.tsx:
```diff
@@ -0,0 +1,51 @@
+        <BugReport isResponsive={false} />
+      </Box>
+
+      <CardActions className={actions}>
```
Card actions should only be in a card

thdk replied:
Is there any other component I can use for this cause? I went looking in the designs to find something similar but didn't see much except for 'NewHostRequest.tsx' which also uses CardActions. But it should be only used within a card indeed.

I'll change it here, and maybe also on NewHostRequest?

thdk replied:
nm I see NewHostRequest itself is being wrapped in a Card component by it's parent.
I've created an Actions component that wraps these buttons when used outside of a Card. If that's too much I can move the styling inside the ErrorFallback component.

lucaslcode commented on app/frontend/src/components/ErrorFallback/ErrorFallback.tsx:
```diff
@@ -0,0 +1,51 @@
+      </Box>
+
+      <CardActions className={actions}>
+        <NavLink exact to={{ pathname: baseRoute }}>
+          <Button variant="outlined">{GO_TO_HOMEPAGE}</Button>
+        </NavLink>
```
1. Use `Link` (make sure it's from `react-router-dom` not mui), `NavLink` is used where you need an "active" class.
2. This isn't valid html (button inside anchor).  use `<Link component={Button} ...`

lucaslcode commented on app/frontend/src/components/ErrorFallback/constants.ts:
```diff
@@ -0,0 +1,5 @@
+export const ERROR_HEADING = "Ooops... that's an error";
+export const ERROR_INFO =
+  "Please use the 'Report a problem' button to help us prevent this error in the future.";
```
```suggestion
export const ERROR_HEADING = "Ooops... that's an unknown error";
export const ERROR_INFO =
  "Please use the 'Report a problem' button and tell us what you were doing to help us prevent this error in the future.";
```

lucaslcode reviewed (changes requested):
Thanks! 👍🏻

thdk committed:
Update app/frontend/src/components/ErrorFallback/constants.ts

Co-authored-by: Lucas Levin <lucas@couchers.org>

thdk committed:
style: review feedback

lucaslcode commented on app/frontend/src/AppRoute.tsx:
```diff
@@ -46,6 +47,8 @@ export default function AppRoute({
 
   const classes = useStyles();
 
+  const RouteContent = () => <ErrorBoundary>{children}</ErrorBoundary>;
```
Sorry, I'm an idiot, components in a render function do need to be memoized. Better still would be moving it into a separate function. That way it doesn't get recreated and also doesn't have the (small) overhead of memoizing.

lucaslcode replied:
Alternatively, I'd also consider moving the error boundary to `App.tsx`, plus that way it will help in case there's an error in the navigation or theme.

thdk replied:
Not sure if we need to move the ErrorBoundary from AppRoute to App level. Instead we can add an additional ErrorBoundary to App.

When a Route throws an error, it will still show the top navigation bar and the error will be reset when user navigates to another page. While if we just have a root ErrorBoundary, all the user will see is the ErrorFallback component.

Does that makes sense?

lucaslcode replied:
Yes good point

lucaslcode commented on app/frontend/src/components/Actions.tsx:
```diff
@@ -0,0 +1,19 @@
+  },
+}));
+
+export default function Actions({ children }: { children: ReactNode }) {
```
Could you also add this to storybook by making an `Actions.stories.tsx` file?

thdk replied:
yes, shall do :)

lucaslcode commented on app/frontend/src/components/ErrorFallback/ErrorFallback.tsx:
```diff
@@ -0,0 +1,47 @@
+      </div>
+
+      <Actions>
+        <Link
+          to={{ pathname: baseRoute }}
+          component={() => <Button variant="outlined">{GO_TO_HOMEPAGE}</Button>}
+        />
```
```suggestion
        <Link
          to={{ pathname: baseRoute }}
          component={Button}
          variant="outlined"
        >
          {GO_TO_HOMEPAGE}
        </Link>
```

lucaslcode replied:
Wait! I think this is also wrong

lucaslcode replied:
```suggestion
        <Button
          to={{ pathname: baseRoute }}
          component={Link}
          variant="outlined"
        >
          {GO_TO_HOMEPAGE}
        </Button>
```
Try that. Check the console for errors

thdk committed:
feat: add fatal error boundary

Since our custom ErrorBoundary needs the context object to display the BugReport button, I have create a variant of the ErrorFallback for fatal errors (aka errors not wrapped yet by anothera ErrorBoundary).

This variant still needs the styling from the theme component. If we really want a root ErrorHandler, then the fallback component should be plain html.

thdk committed:
fix: error boundary should update when location changes

thdk committed:
style: don't wrap ErrorBoundary in another RouteContent component

thdk reviewed (commented):
Somehow I started a review, and I can't find a button to cancel this :(

thdk committed:
fix: click on back to homepage doesn't work for fullscreen routes

thdk committed:
style: move Actions into its own component folder

I prefere all files related to one component (tests, stories, container, ...) to be in one folder using an index file to reexport the component.

thdk committed:
story: add three reusable stories for Button component

Having stories that display a single component (adjustable by controls) is easier so it can be reused by composite stories

thdk committed:
story: add stories for Actions component

thdk committed:
story: fix Outlined variant in Button story

Don't use any. Grrrr! :)

thdk commented:
Fixes #1082

lucaslcode commented on app/frontend/src/components/Actions/Actions.tsx:
```diff
@@ -0,0 +1,19 @@
+    justifyContent: "flex-end",
+    padding: theme.spacing(1),
+    "& :not(:first-child)": {
+      marginLeft: theme.spacing(1),
```
```suggestion
      marginInlineStart: theme.spacing(1),
```
`marginInlineStart` and `marginInlineEnd` (but *not* `marginInline`) have good browser compatibility these days, so we have switched to using them in case we ever want to support RtL languages (though there will probably still be plenty of other stuff to fix up!)

lucaslcode replied:
You could also do `marginBlockStart` instead of `marginTop` if you want, but I don't think it's that important.

lucaslcode commented on app/frontend/src/App.tsx:
```diff
@@ -14,13 +17,17 @@ function App() {
+        <FatalErrorBoundary fallbackRender={() => <ErrorFallback isFatal />}>
+          <ReactQueryClientProvider>
+            <AuthProvider>
+              <ErrorBoundary>
```
Why this one as well as the fatal one? It also seems a bit odd to have our component `ErrorBoundary` but also use the one from the package directly.

lucaslcode reviewed (changes requested):
Looks great! One or two very minor nits

thdk committed:
style: use marginInlineStart

thdk committed:
feat: reuse error boundary component for fatal error

lucaslcode commented on app/frontend/src/components/ErrorBoundary/ErrorBoundary.tsx:
```diff
@@ -1,15 +1,22 @@
+export default function ErrorBoundary({
+  children,
+  ...otherProps
+}: PropsWithChildren<Partial<ErrorBoundaryProps>>) {
```
Doesn't `ErrorBoundaryProps` already contain `children`? So maybe `PropsWithChildren` isn't needed.

thdk replied:
```javascript
declare type ErrorBoundaryProps = ErrorBoundaryPropsWithFallback | ErrorBoundaryPropsWithComponent | ErrorBoundaryPropsWithRender;
declare type ErrorBoundaryState = {
    error: Error | null;
};
declare class ErrorBoundary extends React.Component<React.PropsWithRef<React.PropsWithChildren<ErrorBoundaryProps>>, ErrorBoundaryState> {
```

thdk replied:
These props are also being wrapped by the ErrorBoundary component (the one from react-error-boundary) using `PropsWithChildren`. The type `ErrorBoundaryProps` itself doesn't have the children prop yet.

lucaslcode reviewed (approved):
1 minor thing, if I'm incorrect feel free to merge.
